File: src/CsharpClient/QuixStreams.State/Storage/InMemoryStorage.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace QuixStreams.State.Storage
{
    /// <summary>
    /// Basic non-thread safe in-memory storage implementing <see cref="IStateStorage"/> interface.
    /// </summary>
    public class InMemoryStorage : IStateStorage
    {
        /// <summary>
        /// Represents the in-memory state holding the key-value pairs.
        /// </summary>
        private readonly IDictionary<string, byte[]> inMemoryState = new Dictionary<string, byte[]>();
        
        private readonly string keyPrefix;
        private readonly string storageName;
        
        private const char Separator = '/';
        
        /// <summary>
        /// Instantiates a new instance of <see cref="InMemoryStorage"/>
        /// </summary>
        /// <param name="streamId">Stream id of the storage</param>
        /// <param name="stateName">Stream id of the storage</param>
        public static InMemoryStorage GetStateStorage(string streamId, string stateName)
        {
            if (string.IsNullOrEmpty(streamId) || string.IsNullOrEmpty(stateName))
            {
                throw new ArgumentException($"{nameof(streamId)} and {nameof(stateName)} cannot be null or empty.");
            }
            
            var storageName = $"{streamId}{Separator}{stateName}";
            return new InMemoryStorage(storageName);
        }
        
        /// <summary>
        /// Instantiates a new instance of <see cref="InMemoryStorage"/>
        /// </summary>
        /// <param name="storageName">Name of the storage. Used as a prefix to separate data of other states when they use the same db</param>
        private InMemoryStorage(string storageName)
        {
            if (string.IsNullOrEmpty(storageName))
            {
                throw new ArgumentException($"{nameof(storageName)} cannot be null or empty.");
            }
            
            this.keyPrefix = $"{storageName}{Separator}";
            this.storageName = storageName;
        }
        
        /// <inheritdoc/>
        public Task SaveRaw(string key, byte[] data)
        {
            var prefixedKey = this.keyPrefix + key;
            this.inMemoryState[prefixedKey] = data;
            return Task.CompletedTask;
        }

        /// <inheritdoc/>
        public Task<byte[]> LoadRaw(string key)
        {
            var prefixedKey = this.keyPrefix + key;
            return Task.FromResult(this.inMemoryState[prefixedKey]);
        }

        /// <inheritdoc/>
        public Task RemoveAsync(string key)
        {
            var prefixedKey = this.keyPrefix + key;
            this.inMemoryState.Remove(prefixedKey);
            return Task.CompletedTask;
        }

        /// <inheritdoc/>
        public Task<bool> ContainsKeyAsync(string key)
        {
            var prefixedKey = this.keyPrefix + key;
            var contains = this.inMemoryState.ContainsKey(prefixedKey);
            return Task.FromResult(contains);
        }

        /// <inheritdoc/>
        public Task<string[]> GetAllKeysAsync()
        {
            var keys = inMemoryState.Keys.Select(x => x.Substring(this.keyPrefix.Length)).ToArray();
            return Task.FromResult(keys);
        }

        /// <inheritdoc/>
        public Task ClearAsync()
        {
            this.inMemoryState.Clear();
            return Task.CompletedTask;
        }
        
        /// <inheritdoc/>
        public Task<int> Count()
        {
            return Task.FromResult(this.inMemoryState.Count);
        }

        /// <inheritdoc/>
        public bool IsCaseSensitive => true;
        
        /// <inheritdoc/>
        public void Flush() => throw new NotSupportedException();
        
        /// <inheritdoc/>
        public bool CanPerformTransactions => false;
        
        /// <summary>
        /// Deletes all the states of a stream
        /// </summary>
        /// <param name="streamId">Stream id of the storage</param>
        public static void DeleteStreamStates(string streamId)
        {
            if (string.IsNullOrEmpty(streamId))
            {
                throw new ArgumentException($"{nameof(streamId)} cannot be null or empty.");
            }
            
            var streamStorage = new InMemoryStorage(storageName: streamId);
            streamStorage.Clear();
        }

        /// <inheritdoc />
        public void Dispose()
        {
            // Nothing to dispose
        }
    }
}

File: src/PythonClient/src/quixstreams/statestorages/inmemorystorage.py
import threading
from typing import Dict, List

from ..statestorages.istatestorage import IStateStorage


class InMemoryStorage(IStateStorage):
    """Basic non-thread safe in-memory storage implementing IStateStorage protocol."""
    # IS IT REALLY NON-THREAD SAFE??

    def __init__(self):
        self._in_memory_state: Dict[str, bytes] = {}
        self._sub_states: Dict[str, 'InMemoryStorage'] = {}
        self._sub_state_lock = threading.Lock()

    def save_raw(self, key: str, data: bytes) -> None:
        """Save raw data into the key"""
        self._in_memory_state[key] = data

    def load_raw(self, key: str) -> bytes:
        """Load raw data from the key"""
        return self._in_memory_state[key]

    def remove(self, key: str) -> None:
        """Remove key from the storage"""
        self._in_memory_state.pop(key, None)

    def contains_key(self, key: str) -> bool:
        """Check if storage contains key"""
        return key in self._in_memory_state

    def get_all_keys(self) -> List[str]:
        """Get list of all keys in the storage"""
        return list(self._in_memory_state.keys())

    def clear(self) -> None:
        """Clear the storage / remove all keys from the storage"""
        self._in_memory_state.clear()

    def count(self) -> int:
        """Returns the number of keys in storage"""
        return len(self._in_memory_state)

    @property
    def is_case_sensitive(self) -> bool:
        """Returns whether the storage is case-sensitive"""
        return True

    def get_or_create_sub_storage(self, sub_storage_name: str) -> 'InMemoryStorage':
        """Creates or retrieves the existing storage under this in hierarchy."""
        with self._sub_state_lock:
            if sub_storage_name in self._sub_states:
                return self._sub_states[sub_storage_name]

            storage = InMemoryStorage()
            self._sub_states[sub_storage_name] = storage
            return storage

    def delete_sub_storage(self, sub_storage_name: str) -> bool:
        """Deletes a storage under this in hierarchy."""
        return self._sub_states.pop(sub_storage_name, None) is not None

    def delete_sub_storages(self) -> int:
        """Deletes the storages under this in hierarchy."""
        count = len(self._sub_states)
        self._sub_states.clear()
        return count

    def get_sub_storages(self) -> List[str]:
        """Gets the storages under this in hierarchy."""
        return list(self._sub_states.keys())
